planner: update document about non-prepared plan cache (#14654)
pingcap/docs: sql-prepared-plan-cache.md

- The query accesses partitioned tables or temporary tables, or a table that contains generated columns.
- The query contains non-correlated sub-queries, such as `SELECT * FROM t1 WHERE t1.a > (SELECT 1 FROM t2 WHERE t2.b < 1)`.
- The query contains correlated sub-queries with `PhysicalApply` operators in the execution plan, such as `SELECT * FROM t1 WHERE t1.a > (SELECT a FROM t2 WHERE t1.b > t2.b)`.

> - The query contains the `ignore_plan_cache` or `set_var` hint, such as `select /*+ ignore_plan_cache() */ * from t` or `select /*+ set_var(max_execution_time=1) */ * from t`.

Reviewer: ```suggestion
- The query contains the `ignore_plan_cache` or `set_var` hint, such as `SELECT /*+ ignore_plan_cache() */ * FROM t` or `SELECT /*+ set_var(max_execution_time=1) */ * FROM t`.
```

- The query contains parameters for comparing `int` and `string`, such as `c_int >= ?` or `c_int in (?, ?)`, in which `?` indicates the string type, such as `set @x='123'`. To ensure that the query result is compatible with MySQL, parameters need to be adjusted in each query, so such queries are not cached.
- The plan attempts to access `TiFlash`.
- In most cases, the plan that contains `TableDual` is not cached, unless the current `Prepare` statement does not have parameters.

> - The query accessing TiDB system views, such as `information_schema.columns`, and it is not recommended to use `Prepare` and `Execute` statements to access system views. 
> 
> TiDB has a limitation on the number of `?`, if a query has more than 65535 `?`, the error `Prepared statement contains too many placeholders` will be returned.

Reviewer: ```suggestion
- The query accesses TiDB system views, such as `information_schema.columns`. It is not recommended to use `Prepare` and `Execute` statements to access system views.

TiDB has a limitation on the number of `?` in a query. If a query contains more than 65535 `?`, an error `Prepared statement contains too many placeholders` is reported.
```
